cli/pkg/release/parsing.go

```diff
@@ -20,80 +20,275 @@ import (
-	environments cli_utils.RepeatedString
-	manifests    cli_utils.RepeatedString
+// unless you're working on the readArgs and parseArgs functions, you probably don't need this type, see ReleaseParameters instead
+type commandLineArguments struct {
+	application      cli_utils.RepeatedString // code-simplifying hack: we use RepeatingString for application even though it's not meant to be repeated so that we can raise and error when it's repeated more or less than once
+	environments     cli_utils.RepeatedString
+	manifests        cli_utils.RepeatedString
+	team             cli_utils.RepeatedString // same hack as application field here
+	sourceCommitId   cli_utils.RepeatedString // same hack as application field here
+	previousCommitId cli_utils.RepeatedString // same hack as application field here
+	sourceAuthor     cli_utils.RepeatedString // same hack as application field here
+	sourceMessage    cli_utils.RepeatedString // same hack as application field here
+	version          cli_utils.RepeatedString // same hack as application field here
+	displayVersion   cli_utils.RepeatedString // same hack as application field here
+	skipSignatures   bool
+	signatures       cli_utils.RepeatedString // same hack as application field here
 }
 
-func parseArgs(args []string) (*cmdArguments, error) {
-	cmdArgs := cmdArguments{}
+// checks whether every --environment arg is matched with a --manifest arg
+func environmentsManifestsPaired(args []string) (result bool, message string) {
+	for i, arg := range args {
+		if arg == "--environment" {
+			nextIndex := i + 2
+			if nextIndex >= len(args) || args[nextIndex] != "--manifest" {
+				return false, "all --environment args must have a --manifest arg set immediately afterwards"
+			}
+		}
+		if arg == "--manifest" {
+			prevIndex := i - 2
+			if prevIndex < 0 || args[prevIndex] != "--environment" {
+				return false, "all --manifest args must be set immediately after an --environment arg"
+			}
+		}
+	}
+	return true, ""
+}
+
+// checks whether every --manifest arg is matched with a --signature arg
+func manifestsSignaturesPaired(args []string) (result bool, message string) {
+	for i, arg := range args {
+		if arg == "--manifest" {
+			nextIndex := i + 2
+			if nextIndex >= len(args) || args[nextIndex] != "--signature" {
+				return false, "all --manifest args must have a --signature arg set immediately afterwards, unless --skip_signatures is set"
+			}
+		}
+		if arg == "--signature" {
+			prevIndex := i - 2
+			if prevIndex < 0 || args[prevIndex] != "--manifest" {
+				return false, "all --signature args must be set immediately after a --manifest arg"
+			}
+		}
+	}
+	return true, ""
+}
+
+func argsValid(cmdArgs *commandLineArguments) (result bool, message string) {
+	var commitIDRegex = regexp.MustCompile(`^[0-9a-fA-F]{40}$`)
+	isCommitID := func(s string) bool {
+		return commitIDRegex.MatchString(s)
+	}
+
+	var authorIDRegex = regexp.MustCompile(`^[^<\n]+( <[^@\n]+@[^>\n]+>)?$`)
```

Comment:
since we are using the same regexp as elsewhere, we need to make sure that they always match.
```suggestion
        // this regexp needs to be the same as in ... <link to file>
	var authorIDRegex = regexp.MustCompile(`^[^<\n]+( <[^@\n]+@[^>\n]+>)?$`)
```
And the same at the other place (release.go).